File: sdks/python/nexuszero/privacy.py
"""
Adaptive Privacy Morphing (APM) Engine for NexusZero Protocol.

Implements the 6-level privacy spectrum:
- Level 0: Transparent - Public blockchain parity
- Level 1: Pseudonymous - Address obfuscation
- Level 2: Confidential - Encrypted amounts
- Level 3: Private - Full transaction privacy
- Level 4: Anonymous - Unlinkable transactions
- Level 5: Sovereign - Maximum privacy, ZK everything
"""

from dataclasses import dataclass, field
from enum import IntEnum
from typing import Optional


class PrivacyLevel(IntEnum):
    """The 6-level privacy spectrum."""

    TRANSPARENT = 0  # Public blockchain parity
    PSEUDONYMOUS = 1  # Address obfuscation
    CONFIDENTIAL = 2  # Encrypted amounts
    PRIVATE = 3  # Full transaction privacy
    ANONYMOUS = 4  # Unlinkable transactions
    SOVEREIGN = 5  # Maximum privacy, ZK everything

    @property
    def description(self) -> str:
        """Human-readable description of the privacy level."""
        descriptions = {
            0: "Transparent: Public blockchain parity, no privacy enhancements",
            1: "Pseudonymous: Address obfuscation with basic decoys",
            2: "Confidential: Encrypted amounts, visible addresses",
            3: "Private: Full transaction privacy with ZK proofs",
            4: "Anonymous: Unlinkable transactions, large anonymity set",
            5: "Sovereign: Maximum privacy, quantum-resistant ZK proofs",
        }
        return descriptions[self.value]

    @property
    def security_bits(self) -> int:
        """Security level in bits."""
        security = {0: 0, 1: 80, 2: 128, 3: 192, 4: 256, 5: 256}
        return security[self.value]

    @property
    def estimated_proof_time_ms(self) -> int:
        """Estimated proof generation time in milliseconds."""
        times = {0: 0, 1: 50, 2: 100, 3: 250, 4: 500, 5: 1000}
        return times[self.value]


@dataclass
class PrivacyParameters:
    """Configuration parameters for a specific privacy level."""

    lattice_n: int  # Lattice dimension
    modulus_q: int  # Modulus
    sigma: float  # Error distribution standard deviation
    security_bits: int
    proof_strategy: str
    anonymity_set_size: Optional[int] = None
    decoy_count: Optional[int] = None


@dataclass
class TransactionContext:
    """Context for privacy level recommendation."""

    value_usd: float
    requires_compliance: bool = False
    preferred_level: Optional[int] = None
    risk_score: float = 0.0
    jurisdiction: str = "US"
    counterparty_known: bool = False


@dataclass
class PrivacyRecommendation:
    """Recommendation from the APM engine."""

    level: PrivacyLevel
    parameters: PrivacyParameters
    reasons: list[str] = field(default_factory=list)
    estimated_proof_time_ms: int = 0
    estimated_cost_gas: int = 0


class PrivacyEngine:
    """
    Adaptive Privacy Morphing (APM) Engine.

    Provides dynamic privacy level management and recommendations
    based on transaction context, regulatory requirements, and user preferences.
    """

    # Default parameters for each privacy level
    LEVEL_PARAMETERS: dict[int, PrivacyParameters] = {
        0: PrivacyParameters(
            lattice_n=0,
            modulus_q=0,
            sigma=0.0,
            security_bits=0,
            proof_strategy="none",
        ),
        1: PrivacyParameters(
            lattice_n=256,
            modulus_q=12289,
            sigma=3.2,
            security_bits=80,
            proof_strategy="bulletproofs",
            decoy_count=3,
        ),
        2: PrivacyParameters(
            lattice_n=512,
            modulus_q=12289,
            sigma=3.2,
            security_bits=128,
            proof_strategy="bulletproofs",
            decoy_count=7,
        ),
        3: PrivacyParameters(
            lattice_n=1024,
            modulus_q=40961,
            sigma=3.2,
            security_bits=192,
            proof_strategy="quantum_lattice_pkc",
            anonymity_set_size=16,
            decoy_count=15,
        ),
        4: PrivacyParameters(
            lattice_n=2048,
            modulus_q=65537,
            sigma=3.2,
            security_bits=256,
            proof_strategy="quantum_lattice_pkc",
            anonymity_set_size=64,
            decoy_count=31,
        ),
        5: PrivacyParameters(
            lattice_n=4096,
            modulus_q=786433,
            sigma=3.2,
            security_bits=256,
            proof_strategy="hybrid_zksnark_lattice",
            anonymity_set_size=256,
            decoy_count=63,
        ),
    }

    GAS_COSTS: dict[int, int] = {
        0: 21000,
        1: 50000,
        2: 100000,
        3: 200000,
        4: 350000,
        5: 500000,
    }

    def __init__(self) -> None:
        """Initialize the APM engine with default parameters."""
        self.parameters = self.LEVEL_PARAMETERS.copy()

    def get_parameters(self, level: int) -> PrivacyParameters:
        """Get parameters for a specific privacy level."""
        if level not in self.parameters:
            raise ValueError(f"Invalid privacy level: {level}")
        return self.parameters[level]

    def recommend(self, context: TransactionContext) -> PrivacyRecommendation:
        """
        Recommend optimal privacy level based on transaction context.

        Args:
            context: Transaction context including value, compliance needs, etc.

        Returns:
            PrivacyRecommendation with level, parameters, and reasoning.
        """
        recommended_level = 3  # Default to Private
        reasons: list[str] = []
        compliance_cap: int | None = None

        # Regulatory considerations - track as cap to enforce at end
        if context.requires_compliance:
            compliance_cap = 3
            reasons.append("Regulatory compliance limits maximum privacy to Level 3")

        # Transaction value considerations
        if context.value_usd > 100_000.0:
            recommended_level = 5
            reasons.append("Very high-value transaction ($100k+) warrants Sovereign privacy")
        elif context.value_usd > 10_000.0:
            recommended_level = max(recommended_level, 4)
            reasons.append("High-value transaction ($10k+) benefits from Anonymous privacy")

        # User preference
        if context.preferred_level is not None:
            recommended_level = context.preferred_level
            reasons.append(f"User preference: Level {context.preferred_level}")

        # Enforce compliance cap after all other adjustments
        if compliance_cap is not None and recommended_level > compliance_cap:
            recommended_level = compliance_cap

        # Risk score adjustment
        if context.risk_score > 0.7:
            recommended_level = min(recommended_level, 2)
            reasons.append("Elevated risk score (>0.7) reduces maximum privacy")

        # Counterparty considerations
        if context.counterparty_known:
            reasons.append("Known counterparty may allow reduced privacy overhead")

        level = PrivacyLevel(recommended_level)
        params = self.get_parameters(recommended_level)

        return PrivacyRecommendation(
            level=level,
            parameters=params,
            reasons=reasons,
            estimated_proof_time_ms=level.estimated_proof_time_ms,
            estimated_cost_gas=self.GAS_COSTS.get(recommended_level, 0),
        )

    def can_morph(self, from_level: int, to_level: int) -> tuple[bool, str]:
        """
        Check if morphing between privacy levels is possible.

        Args:
            from_level: Current privacy level
            to_level: Target privacy level

        Returns:
            Tuple of (can_morph, reason)
        """
        if from_level < 0 or from_level > 5:
            return False, f"Invalid source level: {from_level}"
        if to_level < 0 or to_level > 5:
            return False, f"Invalid target level: {to_level}"

        # Increasing privacy is always allowed (new proof)
        if to_level >= from_level:
            return True, "Increasing or maintaining privacy level is always allowed"

        # Decreasing privacy requires incremental steps
        return True, "Decreasing privacy must be done incrementally to preserve guarantees"

    def get_morph_path(self, from_level: int, to_level: int) -> list[int]:
        """
        Get the morphing path between privacy levels.

        Args:
            from_level: Current privacy level
            to_level: Target privacy level

        Returns:
            List of levels to transition through
        """
        if to_level > from_level:
            # Increasing privacy - can do in one step
            return [to_level]
        else:
            # Decreasing privacy - must do incrementally
            return list(range(from_level - 1, to_level - 1, -1))

File: sdks/go/nexuszero/privacy.go
package nexuszero

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"time"

	"github.com/google/uuid"
)

// PrivacyParameters holds configuration for a specific privacy level.
type PrivacyParameters struct {
	LatticeN         int     `json:"lattice_n"`
	ModulusQ         uint64  `json:"modulus_q"`
	Sigma            float64 `json:"sigma"`
	SecurityBits     int     `json:"security_bits"`
	ProofStrategy    string  `json:"proof_strategy"`
	AnonymitySetSize *int    `json:"anonymity_set_size,omitempty"`
	DecoyCount       *int    `json:"decoy_count,omitempty"`
}

// TransactionContext holds context for privacy level recommendations.
type TransactionContext struct {
	ValueUSD           float64 `json:"value_usd"`
	RequiresCompliance bool    `json:"requires_compliance"`
	PreferredLevel     *int    `json:"preferred_level,omitempty"`
	RiskScore          float64 `json:"risk_score"`
	Jurisdiction       string  `json:"jurisdiction"`
	CounterpartyKnown  bool    `json:"counterparty_known"`
}

// PrivacyRecommendation is the result of privacy level recommendation.
type PrivacyRecommendation struct {
	Level               PrivacyLevel      `json:"level"`
	Parameters          PrivacyParameters `json:"parameters"`
	Reasons             []string          `json:"reasons"`
	EstimatedProofTimeMs int              `json:"estimated_proof_time_ms"`
	EstimatedCostGas    uint64            `json:"estimated_cost_gas"`
}

// PrivacyEngine provides adaptive privacy morphing functionality.
type PrivacyEngine struct {
	parameters map[PrivacyLevel]PrivacyParameters
	gasCosts   map[PrivacyLevel]uint64
}

// NewPrivacyEngine creates a new privacy engine with default parameters.
func NewPrivacyEngine() *PrivacyEngine {
	decoy3 := 3
	decoy7 := 7
	decoy15 := 15
	decoy31 := 31
	decoy63 := 63
	anon16 := 16
	anon64 := 64
	anon256 := 256

	return &PrivacyEngine{
		parameters: map[PrivacyLevel]PrivacyParameters{
			PrivacyTransparent: {
				LatticeN:      0,
				ModulusQ:      0,
				Sigma:         0.0,
				SecurityBits:  0,
				ProofStrategy: "none",
			},
			PrivacyPseudonymous: {
				LatticeN:      256,
				ModulusQ:      12289,
				Sigma:         3.2,
				SecurityBits:  80,
				ProofStrategy: "bulletproofs",
				DecoyCount:    &decoy3,
			},
			PrivacyConfidential: {
				LatticeN:      512,
				ModulusQ:      12289,
				Sigma:         3.2,
				SecurityBits:  128,
				ProofStrategy: "bulletproofs",
				DecoyCount:    &decoy7,
			},
			PrivacyPrivate: {
				LatticeN:         1024,
				ModulusQ:         40961,
				Sigma:            3.2,
				SecurityBits:     192,
				ProofStrategy:    "quantum_lattice_pkc",
				AnonymitySetSize: &anon16,
				DecoyCount:       &decoy15,
			},
			PrivacyAnonymous: {
				LatticeN:         2048,
				ModulusQ:         65537,
				Sigma:            3.2,
				SecurityBits:     256,
				ProofStrategy:    "quantum_lattice_pkc",
				AnonymitySetSize: &anon64,
				DecoyCount:       &decoy31,
			},
			PrivacySovereign: {
				LatticeN:         4096,
				ModulusQ:         786433,
				Sigma:            3.2,
				SecurityBits:     256,
				ProofStrategy:    "hybrid_zksnark_lattice",
				AnonymitySetSize: &anon256,
				DecoyCount:       &decoy63,
			},
		},
		gasCosts: map[PrivacyLevel]uint64{
			PrivacyTransparent:  21000,
			PrivacyPseudonymous: 50000,
			PrivacyConfidential: 100000,
			PrivacyPrivate:      200000,
			PrivacyAnonymous:    350000,
			PrivacySovereign:    500000,
		},
	}
}

// GetParameters returns the parameters for a specific privacy level.
func (e *PrivacyEngine) GetParameters(level PrivacyLevel) (PrivacyParameters, error) {
	params, ok := e.parameters[level]
	if !ok {
		return PrivacyParameters{}, fmt.Errorf("invalid privacy level: %d", level)
	}
	return params, nil
}

// Recommend returns a privacy level recommendation based on context.
func (e *PrivacyEngine) Recommend(ctx TransactionContext) PrivacyRecommendation {
	recommendedLevel := PrivacyPrivate // Default
	var reasons []string

	// Regulatory considerations
	if ctx.RequiresCompliance {
		if recommendedLevel > PrivacyPrivate {
			recommendedLevel = PrivacyPrivate
		}
		reasons = append(reasons, "Regulatory compliance limits maximum privacy to Level 3")
	}

	// Transaction value considerations
	if ctx.ValueUSD > 10000.0 {
		if recommendedLevel < PrivacyAnonymous {
			recommendedLevel = PrivacyAnonymous
		}
		reasons = append(reasons, "High-value transaction ($10k+) benefits from Anonymous privacy")
	}
	if ctx.ValueUSD > 100000.0 {
		recommendedLevel = PrivacySovereign
		reasons = append(reasons, "Very high-value transaction ($100k+) warrants Sovereign privacy")
	}

	// User preference
	if ctx.PreferredLevel != nil {
		recommendedLevel = PrivacyLevel(*ctx.PreferredLevel)
		reasons = append(reasons, fmt.Sprintf("User preference: Level %d", *ctx.PreferredLevel))
	}

	// Risk score adjustment
	if ctx.RiskScore > 0.7 {
		if recommendedLevel > PrivacyConfidential {
			recommendedLevel = PrivacyConfidential
		}
		reasons = append(reasons, "Elevated risk score (>0.7) reduces maximum privacy")
	}

	params, _ := e.GetParameters(recommendedLevel)

	return PrivacyRecommendation{
		Level:               recommendedLevel,
		Parameters:          params,
		Reasons:             reasons,
		EstimatedProofTimeMs: recommendedLevel.EstimatedProofTimeMs(),
		EstimatedCostGas:    e.gasCosts[recommendedLevel],
	}
}

// CanMorph checks if morphing between privacy levels is possible.
func (e *PrivacyEngine) CanMorph(fromLevel, toLevel PrivacyLevel) (bool, string) {
	if fromLevel > PrivacySovereign || toLevel > PrivacySovereign {
		return false, "Invalid privacy level"
	}

	if toLevel >= fromLevel {
		return true, "Increasing or maintaining privacy level is always allowed"
	}

	return true, "Decreasing privacy must be done incrementally"
}

// GetMorphPath returns the path for morphing between levels.
func (e *PrivacyEngine) GetMorphPath(fromLevel, toLevel PrivacyLevel) []PrivacyLevel {
	if toLevel > fromLevel {
		// Increasing privacy - one step
		return []PrivacyLevel{toLevel}
	}

	// Decreasing privacy - incremental
	var path []PrivacyLevel
	for level := fromLevel - 1; level >= toLevel; level-- {
		path = append(path, level)
	}
	return path
}

// LocalProofResult is the result of local proof generation.
type LocalProofResult struct {
	ProofID          string    `json:"proof_id"`
	ProofData        []byte    `json:"proof_data"`
	PrivacyLevel     int       `json:"privacy_level"`
	GenerationTimeMs int64     `json:"generation_time_ms"`
	QualityScore     float64   `json:"quality_score"`
	GeneratedAt      time.Time `json:"generated_at"`
}

// LocalProofGenerator generates proofs locally (for testing).
type LocalProofGenerator struct {
	engine *PrivacyEngine
}

// NewLocalProofGenerator creates a new local proof generator.
func NewLocalProofGenerator() *LocalProofGenerator {
	return &LocalProofGenerator{
		engine: NewPrivacyEngine(),
	}
}

// Generate generates a proof locally.
func (g *LocalProofGenerator) Generate(data []byte, level PrivacyLevel) (*LocalProofResult, error) {
	start := time.Now()

	// Simulate proof generation
	proofSize := 32 * (int(level) + 1)
	h := sha256.New()
	h.Write(data)
	h.Write([]byte{byte(level)})
	proofHash := h.Sum(nil)

	// Extend proof to expected size
	proofData := make([]byte, 0, proofSize)
	for len(proofData) < proofSize {
		proofData = append(proofData, proofHash...)
	}
	proofData = proofData[:proofSize]

	elapsed := time.Since(start).Milliseconds()

	return &LocalProofResult{
		ProofID:          uuid.New().String(),
		ProofData:        proofData,
		PrivacyLevel:     int(level),
		GenerationTimeMs: max(elapsed, int64(level.EstimatedProofTimeMs())),
		QualityScore:     1.0,
		GeneratedAt:      time.Now(),
	}, nil
}

// LocalProofVerifier verifies proofs locally.
type LocalProofVerifier struct{}

// NewLocalProofVerifier creates a new local proof verifier.
func NewLocalProofVerifier() *LocalProofVerifier {
	return &LocalProofVerifier{}
}

// Verify verifies a proof locally.
func (v *LocalProofVerifier) Verify(proof []byte, level *PrivacyLevel) bool {
	if len(proof) < 32 {
		return false
	}

	if level != nil {
		expectedMinSize := 32 * (int(*level) + 1)
		if len(proof) < expectedMinSize {
			return false
		}
	}

	return true
}

// ProofDataHex returns the proof data as a hex string.
func (r *LocalProofResult) ProofDataHex() string {
	return hex.EncodeToString(r.ProofData)
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}
